// python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. functions.proto
// https://github.com/xainag/numproto
syntax = "proto3";
import "google/protobuf/struct.proto";

message Model {
    string model = 1;
}

message Empty {
    int32 value = 1;
}

message Pair {
   string key = 1;
   string value = 2;
}

message config {
    map<string, google.protobuf.Value> params = 1;
}

message Dictionary {
   repeated Pair pairs = 1;
}

message InitialParams {
    double alpha = 1;
    double index  = 2;
    double isDC = 3;
    int32 dc_index = 4;
    int32 device_index = 5;
    int32 coordinate_per_dc = 6;
    int32 datapoints_per_device = 7;
    int32 randomseed = 8;
    double lambduh = 9;
    bytes stuff = 10;
}

message Note {
    string name = 1;
    string message = 2;
}

service FederatedApp{
    rpc InitializeParams(InitialParams) returns (Empty) {}
    rpc GenerateData(Empty) returns (Empty) {}
    rpc Train(Empty) returns (Model) {} 
    rpc SendModel(Model) returns (Empty) {}
    //rpc ChatStream (Empty) returns (stream Note)
}